File: src/business/rule/engine/serviceimpl/OrderProcessingService.java
package business.rule.engine.serviceimpl;

import business.rule.engine.model.Order;
import business.rule.engine.service.IEmailService;
import business.rule.engine.service.IUserService;
import business.rule.engine.serviceprovider.BusinessRuleProviderFactory;
import business.rule.engine.serviceprovider.IBusinessRuleProvider;

public class OrderProcessingService {
	 private IEnumerable<IBusinessRuleProvider> rules;
     private BusinessRuleProviderFactory businessRuleProviderFactory;

     public OrderProcessingService(IUserService userService, IEmailService emailService)
     {
         this.businessRuleProviderFactory = new BusinessRuleProviderFactory(userService, emailService);
         this.rules = businessRuleProviderFactory.GetProviders();
     }
    
     public void Process(Order order)
     {
         for each (IBusinessRuleProvider rule in rules)
         {
             var (isSatisfied, message) = rule.Process(order);
         }
     }
}

File: src/OrderProcessingService.cs
using Raymaker.RulesEngine.Application.BusinessRuleProviders;
using Raymaker.RulesEngine.Application.Model;
using System.Collections.Generic;

namespace Raymaker.RulesEngine.Application
{
    public class OrderProcessingService
    {
        private readonly IEnumerable<IBusinessRuleProvider> rules;
        private readonly BusinessRuleProviderFactory businessRuleProviderFactory;

        public OrderProcessingService(IUserService userService, IEmailService emailService)
        {
            this.businessRuleProviderFactory = new BusinessRuleProviderFactory(userService, emailService);
            this.rules = businessRuleProviderFactory.GetProviders();
        }

        /// <summary>
        /// Assumes a payment has taken place so we process the order.
        /// </summary>
        public void Process(Order order)
        {
            foreach (IBusinessRuleProvider rule in rules)
            {
                var (isSatisfied, message) = rule.Process(order);
                System.Console.WriteLine(message + ". Result:" + isSatisfied);
            }
        }
    }
}
